src/main/java/software/amazon/awssdk/crt/mqtt/MqttConnection.java

```diff
@@ -144,29 +119,78 @@ public void onSuccess() {
             }
 
             @Override
-            public void onFailure(String reason) {
-                Throwable cause = new MqttException(reason);
-                future.completeExceptionally(cause);
+            public void onFailure(Throwable reason) {
+                future.completeExceptionally(reason);
             }
         };
     }
 
-    public CompletableFuture<Void> connect(String clientId) {
-        return connect(clientId, true, 0);
+    @Override
+    public void onConnectionComplete(int errorCode, boolean sessionPresent) {
+        if (errorCode == 0) {
+            connectionState = ConnectionState.CONNECTED;
+            if (connectAck != null) {
+                connectAck.onSuccess();
+                connectAck = null;
+            }
+        } else {
+            connectionState = ConnectionState.DISCONNECTED;
+            if (connectAck != null) {
+                connectAck.onFailure(new MqttException(errorCode));
+                connectAck = null;
+            }
+        }
+        if (userConnectionCallbacks != null) {
+            userConnectionCallbacks.onConnectionComplete(errorCode, sessionPresent);
+        }
     }
 
-    public CompletableFuture<Void> connect(String clientId, boolean cleanSession) {
-        return connect(clientId, cleanSession, 0);
+    @Override
+    public void onConnectionInterrupted(int errorCode) {
+        connectionState = ConnectionState.DISCONNECTED;
+        if (disconnectAck != null) {
+            if (errorCode == 0) {
+                disconnectAck.onSuccess();
+            } else {
+                disconnectAck.onFailure(new MqttException(errorCode));
+            }
+            disconnectAck = null;
+        }
+        if (userConnectionCallbacks != null) {
+            userConnectionCallbacks.onConnectionInterrupted(errorCode);
+        }
+    }
+
+    @Override
+    public void onConnectionResumed(boolean sessionPresent) {
+        connectionState = ConnectionState.CONNECTED;
+        if (userConnectionCallbacks != null) {
+            userConnectionCallbacks.onConnectionResumed(sessionPresent);
+        }
     }
 
-    public CompletableFuture<Void> connect(String clientId, boolean cleanSession, int keepAliveMs) {
+    public CompletableFuture<Void> connect(String clientId, String endpointUri, int port) {
```

Comment:
Shouldn't the CompletableFuture have `boolean sessionPresent` in it?

Reply:
It could indeed.